metaflow/plugins/cards/card_datastore.py

```diff
@@ -0,0 +1,250 @@
+""" 
+
+"""
+
+from hashlib import sha1
+from io import BytesIO
+import os
+import shutil
+from metaflow.datastore.local_storage import LocalStorage
+
+from .card_decorator import CardDecorator
+from .exception import CardNotPresentException
+
+CARD_DIRECTORY_NAME = "mf.cards"
+TEMP_DIR_NAME = "metaflow_card_cache"
+NUM_SHORT_HASH_CHARS = 5
+
+
+def stepname_from_card_id(card_id, flow):
+    for step in flow._steps:
+        for deco in step.decorators:
+            if isinstance(deco, CardDecorator):
+                if deco.attributes["id"] == card_id:
+                    return step.name
+    return None
+
+
+class CardPathBuilder(object):
+    @classmethod
+    def path_spec_resolver(cls, pathspec):
+        run_id, step_name, task_id = None, None, None
+        splits = pathspec.split("/")
+        if len(splits) == 1:  # only flowname mentioned
+            return splits[0], run_id, step_name, task_id
+        elif len(splits) == 2:  # flowname , runid mentioned
+            return splits[0], splits[1], step_name, task_id
+        elif len(splits) == 3:  # flowname , runid , stepname
+            return splits[0], splits[1], splits[2], task_id
+        elif len(splits) == 4:  # flowname ,runid ,stepname , taskid
+            return splits[0], splits[1], splits[2], splits[3]
+
+    @classmethod
+    def make_path(cls, sysroot, flow_name, run_id=None, task_id=None, pathspec=None):
+        if sysroot is None:
+            return None
+
+        if pathspec is not None:
+            flow_name, run_id, step_name, task_id = cls.path_spec_resolver(pathspec)
+
+        if flow_name is None:
```

Comment:
Oh but it can!. The `path_spec_resolver` takes a string and returns a tuple with 4 objects. Based on the resolved pathspec it will set the non resolved ids to None. 

Example : `Flowname/runid` will return `(flow_name, run_id, None, None)`

Reply:
right, it returns none but never for the first element of the tuple (flow_name). Your example does not contradict this point.

Reply:
Apologies. I misread and commented. Yes you are right. It doesn't return None for Flowname.